src/streamer.ts

```diff
@@ -28,39 +26,82 @@ export async function startStream(
     }
 
     if (blockHeights.length === 0) {
+      // Throttling when there are no new blocks
       await sleep(2000);
       continue;
     }
 
-    const messages = await Promise.all(
-      blockHeights.map(blockHeight => fetchStreamerMessage(s3Client, config.s3BucketName, blockHeight)));
-    for (let streamerMessage of messages) {
-      // check if we have `lastProcessedBlockHash` (might be not set only on start)
-      // compare lastProcessedBlockHash` with `streamerMessage.block.header.prevHash` of the current
-      // block (ensure we never skip blocks even if there is some incident on Lake Indexer side)
-      // retrieve the data from S3 if hashes don't match and repeat the main loop step
-      if (
-        lastProcessedBlockHash &&
-        lastProcessedBlockHash !== streamerMessage.block.header.prevHash
-      ) {
-        console.log(
-          "The hash of the last processed block doesn't match the prevHash of the new one. Refetching the data from S3 in 200ms",
-          lastProcessedBlockHash,
-          streamerMessage.block.header.prevHash,
-        );
-        await sleep(200);
-        break;
+    yield blockHeights.map(blockHeight => fetchStreamerMessage(s3Client, config.s3BucketName, blockHeight));
+    startBlockHeight = Math.max.apply(Math, blockHeights) + 1;
+  }
+}
+
+async function* fetchAhead<T>(seq: AsyncIterable<T>, stepsAhead = 10): AsyncIterableIterator<T> {
+  let queue = [];
+  while (true) {
+    while (queue.length < stepsAhead) {
+      queue.push(seq[Symbol.asyncIterator]().next());
+    }
+
+    const { value, done } = await queue.shift();
+    if (done) return;
+    yield value;
+  }
+}
+
+export async function* stream(
+  config: LakeConfig
+): AsyncIterableIterator<StreamerMessage> {
+  const s3Client = new S3Client({ region: config.s3RegionName });
+
+  let lastProcessedBlockHash: string;
+  let startBlockHeight = config.startBlockHeight;
+  
+  while (true) {
+    try {
+      for await (let promises of fetchAhead(batchStream({ ...config, startBlockHeight }))) {
```

Comment:
Pass `blocksPreloadPoolSize` as stepsAhead into fetchAhead function

Reply:
@frol this is different number however

this fetches ahead given number of batches, each of them being `blocksPreloadPoolSize`